manager/volume.go

```diff
@@ -614,7 +617,58 @@ func (m *VolumeManager) Expand(volumeName string, size int64) (v *longhorn.Volum
 	}
 
 	if v.Spec.Size > size {
-		return nil, fmt.Errorf("cannot expand volume %v with current size %v to a smaller size %v", v.Name, v.Spec.Size, size)
+		return nil, fmt.Errorf("longhorn volume %v spec size %v is greater than requested Persistent Volume Claim size %v", v.Name, v.Spec.Size, size)
+	}
+
+	kubernetesStatus := &v.Status.KubernetesStatus
+	if kubernetesStatus.PVCName != "" && kubernetesStatus.LastPVCRefAt == "" {
+		pvc, err := m.ds.GetPersistentVolumeClaim(kubernetesStatus.Namespace, kubernetesStatus.PVCName)
+		if err != nil {
+			return nil, err
+		}
+
+		requestedSize := resource.MustParse(strconv.FormatInt(size, 10))
+
+		// TODO: Should check for pvc.Spec.Resources.Requests.Storage() here, once upgrade API to v0.18.x.
+		pvcSpecValue, ok := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
+		if !ok {
+			return nil, fmt.Errorf("cannot get request storage")
+		}
+		// TODO: Should check for pvc.Status.CapacityStorage() here, once upgrade API to v0.18.x.
+		pvcStatusValue, ok := pvc.Status.Capacity[corev1.ResourceStorage]
+		if !ok {
+			return nil, fmt.Errorf("cannot get capacity storage")
+		}
+
+		if pvcSpecValue.Cmp(requestedSize) == 0 && pvcStatusValue.Cmp(requestedSize) == 0 && v.Spec.Size == size {
+			// lhv size does not reflect pvc/pv size
+			logrus.Infof("Persistent Volume Claim %v expansion is not necessary since current size %v == %v", v.Name, v.Spec.Size, size)
+			return v, nil
+		} else if pvcSpecValue.Cmp(requestedSize) > 0 {
+			// this is possible due to user can patch lvh, but not patch pv and pvc
+			return nil, fmt.Errorf("persistent volume claim %v spec size %v is greater than requested persistent volume claim size %v", kubernetesStatus.PVCName, pvcSpecValue.Value(), requestedSize.Value())
+		} else if pvcStatusValue.Cmp(requestedSize) > 0 {
+			// this is possible due to user manually created pvc
+			return nil, fmt.Errorf("persistent volume claim %v status size %v is greater than requested persistent volume claim size %v", kubernetesStatus.PVCName, pvcStatusValue.Value(), requestedSize.Value())
```

Comment:
Remove all of these conditions, non of these are our concern.

The only concerns that require updates are:
if pvc.size < requestedSize: -> update PVC
if volume.size < requestedSize: -> update volume

Reply:
Same as above. Some scenarios I found if user edit LHV/PVC manifest manually, it will cause LVH and PVC/PV spec/status size inconsistent so the purpose of it is to prevent it from happen, if it is not our concern, I am ok to remove it.

Reply:
I think there might be a misunderstanding on the conceptual level here:
- A longhorn volume represents a PV
- A PVC is a claim for a PV (the right to exclusively use that PV)

It's valid if a PV.size > PVC.size, for example in a non dynamic provisioning scenario:
- cluster has a PV of size 10GB defined
- user creates a PVC with requested size 1GB, the cluster will associate that PVC (1GB) with the above PV (10GB)

From the above we can see that it should also be valid if Longhorn.Volume.size (actual capacity) > PVC.size (requested capacity)

Reply:
One quick question, should we continue trying to expand Longhorn volume even if the requested size is smaller than the PVC spec size?

Reply:
Yeah I think so, this case should only happen in case of restoration of a smaller volume backup 

we could also do size = max(requestedSize, PVC.requestedSize) 

To cover the restore case but it would actually need to happen as part of the controller to reconcile v.size with the PVC.size.

So we can ignore the controller part for now and do the max(x,y) approach.

------------

For the normal part in which the Cai driver will be called by reconciliation of PVC.size > pv.size this would be covered already, but doing the max approach is beneficial since it allows the user via an expand call to bring a restored volume back in line with the old PVC/PV 

On the CSi expand call since we return the volume.size which will end up in the PV. Kubernetes should end up call expand again while reconciling PVC.requested against the PV.size in the future.

--------

tldr doing size = max(pvc.requestedSize, requestedSize) should have no downaide

Reply:
Do you mean restore/create a volume with small size with existing old PVC? The actual storage size being smaller than the claim desired size is really weird. And doing max is tricky and misunderstanding. e.g., If the requested new size is smaller than both volume spec size and PVC spec size, Longhorn will return success for this expansion call. In the user's perspective, PVC **or/and** volume somehow can be "expanded" to a smaller size since there is no error.

Reply:
updated